File: web/lib/teams.ts
import teamsData from "../../data/teams.json";

export type TeamRow = {
  slug: string;
  abbrev: string;
  name: string;
  conference: string;
  division: string;
  division_order: number;
};

export function getTeams(): TeamRow[] {
  return teamsData as TeamRow[];
}

export type TeamNameLexicon = {
  /** Full franchise names, longest first (e.g. "Washington Commanders"). */
  fullNames: string[];
  /** Lowercased nicknames (commanders, cowboys, jets). */
  nicknames: Set<string>;
  /** Lowercased city phrases, longest first ("new york", "washington"). */
  cities: string[];
};

let teamNameLexicon: TeamNameLexicon | null = null;

/** Franchise tokens used to keep team names from being chipped as players. */
export function getTeamNameLexicon(): TeamNameLexicon {
  if (teamNameLexicon) return teamNameLexicon;
  const teams = getTeams();
  const fullNames = teams
    .map((t) => t.name)
    .sort((a, b) => b.length - a.length);
  const nicknames = new Set<string>();
  const citySet = new Set<string>();
  for (const t of teams) {
    const parts = t.name.split(/\s+/).filter(Boolean);
    const nick = parts[parts.length - 1];
    if (nick) nicknames.add(nick.toLowerCase());
    if (parts.length > 1) {
      citySet.add(parts.slice(0, -1).join(" ").toLowerCase());
    }
  }
  teamNameLexicon = {
    fullNames,
    nicknames,
    cities: [...citySet].sort((a, b) => b.length - a.length),
  };
  return teamNameLexicon;
}

export function getDivisionGroups(): Record<string, TeamRow[]> {
  const order = { East: 0, North: 1, South: 2, West: 3 };
  const teams = teamsData as TeamRow[];
  const grouped: Record<string, TeamRow[]> = {};
  for (const t of [...teams].sort(
    (a, b) =>
      a.conference.localeCompare(b.conference) ||
      order[a.division as keyof typeof order] -
        order[b.division as keyof typeof order] ||
      a.division_order - b.division_order
  )) {
    const key = `${t.conference} ${t.division}`;
    (grouped[key] ??= []).push(t);
  }
  return grouped;
}

File: pipeline/src/teams.py
import json
from dataclasses import dataclass
from pathlib import Path

DATA_PATH = Path(__file__).resolve().parents[2] / "data" / "teams.json"


@dataclass(frozen=True)
class Team:
    slug: str
    abbrev: str
    name: str
    conference: str
    division: str
    division_order: int
    narrative_tier: str
    reddit: str
    keywords: tuple[str, ...]


def load_teams() -> list[Team]:
    raw = json.loads(DATA_PATH.read_text(encoding="utf-8"))
    teams: list[Team] = []
    for row in raw:
        name = row["name"]
        abbrev = row["abbrev"].lower()
        nick = name.split()[-1].lower()
        teams.append(
            Team(
                slug=row["slug"],
                abbrev=abbrev,
                name=name,
                conference=row["conference"],
                division=row["division"],
                division_order=row["division_order"],
                narrative_tier=row["tier"],
                reddit=row["reddit"],
                keywords=(abbrev, nick, name.lower(), row["slug"].replace("-", " ")),
            )
        )
    return teams


def division_groups(teams: list[Team]) -> dict[str, list[Team]]:
    order = {"East": 0, "North": 1, "South": 2, "West": 3}
    grouped: dict[str, list[Team]] = {}
    for t in sorted(teams, key=lambda x: (x.conference, order[x.division], x.division_order)):
        key = f"{t.conference} {t.division}"
        grouped.setdefault(key, []).append(t)
    return grouped
